docs/type_hints.rst

```diff
@@ -0,0 +1,100 @@
+
+There may be cases where some functionality of RDFLib may work perfectly well with values of types that are excluded by the type hints, but if these additional types violate the relevant standards we will consider the correct type hints to be those that exclude values of these types.
+
+Public Type Aliases
+=============================
+In python, type hints are specified in annotations. Type hints are different from type aliases which are normal python variables that are not intended to provide runtime utility and are instead intended for use in static type checking.
+
+For clarify, the following is an example of a function ``foo`` with type hints:
+
+.. code-block:: python
+	
+    def foo(a: int) -> int:
+        return a + 1
+
+While the following is an example of a type alias ``Bar``:
+
+.. code-block:: python
+
+    from typing import Tuple
+
+    Bar = Tuple[int, str]
+
+RDFLib will provide public type aliases under the ``rdflib.typing`` package, for example, ``rdflib.typing.Triple``, ``rdflib.typing.Quad``. Type aliases in the rest of RDFLib should be private (i.e. being with an underscore).
+
+Versioning, Compatibility and Stability
+=======================================
+
+RDFLib attempts to adhere to `semver 2.0 <https://semver.org/spec/v2.0.0.html>`_ which is concerned with the public API of software.
+
+Ignoring type hints, the public API of RDFLib exists implicitly as a consequence of the code of RDFLib and the actual behaviour this entails, the relevant standards that RDFLib is trying to implement, and the documentation of RDFLib, with some interplay between all three of these. RDFLib's public API includes public type aliases, as these are normal python variables and not annotations.
+
+Type hints attempt to formally document RDFLib's implicitly defined public API in a machine-readable fashion as accurately and correctly as possible within the framework outline earlier in this document.
+
+Type hints do not change the functionality of RDFLib, nor do they affect the runtime API of RDFLib. In this way then, they are somewhat outside of the scope of semver, however, they still have an impact on the users of RDFLib, even if this impact is not at runtime, but during development. This necessitates some clarity as to what users of RDFLib should expect regarding type hints in RDFLib releases.
+
+Changes to type hints can broadly be classified as follow:
+
+* **Type Addition**: Adding new type hints to existing code.
+* **Type Narrowing**: Changing type hints of existing code to be narrower. The line between type additions and narrowing is somewhat blurry, as un-annotated code is usually treated the same as if it was annotated with `typing.Any`, so type addition is a form of type narrowing.
```

Comment:
I think (but won't assert) "Type Narrowing" can have a subtly different meaning, per the discussion flowing from [this comment](https://github.com/python/mypy/issues/3004#issuecomment-980470608).  Would "Type Refinement" work for you?

Reply:
changed to "Type Refinement", also added an example.